modules/analysis/src/model.rs

```diff
@@ -0,0 +1,193 @@
+pub struct AdvisoryGraphSummary {
+    pub purl: String,
+    pub name: String,
+    pub published: String,
+    pub document_id: String,
+    pub product_name: String,
+    pub product_version: String,
+
+    pub ancestors: Vec<PackageNode>,
+}
+impl AdvisoryGraphSummary {
+    pub fn new(
+        purl: String,
+        name: String,
+        published: String,
+        document_id: String,
+        product_name: String,
+        product_version: String,
+        ancestors: Vec<PackageNode>,
+    ) -> AdvisoryGraphSummary {
+        AdvisoryGraphSummary {
+            purl,
+            name,
+            published,
+            document_id,
+            product_name,
+            product_version,
+            ancestors,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, ToSchema)]
+pub struct AdvisoryDepSummary {
+    pub purl: String,
+    pub name: String,
+    pub published: String,
+    pub document_id: String,
+    pub product_name: String,
+    pub product_version: String,
+    pub deps: Option<Vec<DepNode>>,
+}
+impl crate::model::AdvisoryDepSummary {
+    pub fn new(
+        purl: String,
+        name: String,
+        published: String,
+        document_id: String,
+        product_name: String,
+        product_version: String,
+        contains: Vec<DepNode>,
+    ) -> crate::model::AdvisoryDepSummary {
+        crate::model::AdvisoryDepSummary {
+            purl,
+            name,
+            published,
+            document_id,
+            product_name,
+            product_version,
+            deps: Some(contains),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct GraphMap {
+    map: HashMap<String, Graph<PackageNode, Relationship, petgraph::Directed>>,
+}
+
+// Static variables for the singleton instance
+static mut INSTANCE: Option<Arc<RwLock<GraphMap>>> = None;
+static INIT: Once = Once::new();
+
+impl GraphMap {
+    // Create a new instance of GraphMap
+    pub fn new() -> Self {
+        GraphMap {
+            map: HashMap::new(),
+        }
+    }
+
+    // Check if the map contains a key
+    pub fn contains_key(&self, key: &str) -> bool {
+        self.map.contains_key(key)
+    }
+
+    // Get the number of graphs in the map
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    // Check if the map is empty
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    // Add a new graph with the given key (write access)
+    pub fn insert(
+        &mut self,
+        key: String,
+        graph: Graph<PackageNode, Relationship, petgraph::Directed>,
+    ) {
+        self.map.insert(key, graph);
+    }
+
+    // Retrieve a reference to a graph by its key (read access)
+    pub fn get(&self, key: &str) -> Option<&Graph<PackageNode, Relationship, petgraph::Directed>> {
+        self.map.get(key)
+    }
+
+    // Get the singleton instance of GraphMap
+    pub fn get_instance() -> Arc<RwLock<GraphMap>> {
+        INIT.call_once(|| unsafe {
```

Comment:
The unsafe gives me the heeby-jeebies.

Reply:
me too - let me see if I can refactor that out